### PR TITLE
Support for JSONConverter in sink connector

### PR DESCRIPTION
**Pull request by tomblench**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

See #62 

## Approach

Support JSONConverter by expecting values from kafka sink tasks to either be a java `Map` or a kafka `Struct`.

- If `Map`, pass through to `batchWrite` directly
- If `Struct`, convert to `Map` using new `StructToMapConverter`

Add documentation in README

## Schema & API Changes

- "No change"

## Security and Privacy

- "No change"

## Testing

See added `StructToMapConverterTests`

## Monitoring and Logging

Added `slf4j-simple` to view log output when running tests

### CODE REVIEW

File: src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java [hunk 1]
```diff
@@ -116,6 +112,7 @@ public static JSONArray batchWrite(Map<String, String> props, JSONArray data)
 				result.put(jsonResult);
 			}
 		} catch (Exception e) {
+			LOG.error("Exception caught in batchWrite()", e);
```
Comment (tomblench) [inline src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java:115]: This may need to be revisited in another PR - the worrying thing is that we were just swallowing exceptions from the cloudant client which I had manage to trigger with a misconfigured test.

File: README.md [hunk 2]
```diff
@@ -29,16 +29,22 @@ the [Kafka Connector documentation](http://docs.confluent.io/3.0.1/connect/userg
...
-Assume these settings in a file `connect-standalone.properties` or `connect-distributed.properties`.
+Usually the kafka distribution defaults (`connect-(standalone|distributed).properties`) are as follows:
+```
+key.converter=org.apache.kafka.connect.json.JsonConverter
+value.converter=org.apache.kafka.connect.json.JsonConverter
+key.converter.schemas.enable=true
+value.converter.schemas.enable=true
+```
+
+#### Converter configuration: sink connector
```
Comment (tomblench) [inline README.md:43]: source connector converter needs covering when we do the PR for that work - my intention is that we support JsonConverter on source and sink which simplifies things (as mentioned above it's the default anyway so no need to explicitly set it in config)

File: src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceAndSinkTest.java [hunk 3]
```diff
@@ -88,13 +92,17 @@ public void testReplicateAll() throws Exception {
             // - no offset
             for (SourceRecord record : records) {
 
+                // source task returns strings but sink task expects structs or maps
+                // in a real kafka instance this would be fixed by using appropriate converters
+                Map recordValue = gson.fromJson((String)record.value(), Map.class);
```
Comment (tomblench) [inline src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceAndSinkTest.java:97]: This is a bit awkward but is needed until we do the PR to support JsonConverter in the source connector - at which point this line can be removed since the source will return us a `Map`

File: src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceTaskTest.java [hunk 4]
```diff
@@ -60,7 +63,8 @@ protected void setUp() throws Exception {
         data = new JSONArray(tokener);
 
         // Load data into the source database (create if it does not exist)
-        JavaCloudantUtil.batchWrite(sourceProperties, data);
+        JavaCloudantUtil.batchWrite(sourceProperties,
```
Comment (tomblench) [inline src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceTaskTest.java:66]: There are a few instances of this awkward mapping from `org.json` in test code - I've tried to make the changes as minimal as possible but would love to get rid of that library altogether at a later date.

File: src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceAndSinkTest.java (hunk 3 above)
Comment (tomblench) [reply src/test/java/com/ibm/cloudant/kafka/connect/CloudantSourceAndSinkTest.java:97]: @ricellis your idea of enabling schemas here (to return a struct/map instead of a string) didn't work because it caused various limits to be exceeded (memory, http request size), because the test payloads are complex resulting in a huge inline schema per doc.

File: src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java (hunk 1 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java:115]: Agreed, there needs to be a separate look at error handling to conform to the behaviours of the built-in `errors.tolerance=all` and `none` flags. (`all` implies silently ignore bad messages so I guess that's all we have right now!)

**Review by ricellis — CHANGES_REQUESTED** [review]
Looks good on the whole, just a few minor suggestions.

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java [hunk 5]
```diff
@@ -48,8 +45,10 @@ public class CloudantSinkTask extends SinkTask {
 	private int taskNumber;
 	public static String guid_schema = null;
 	private Boolean replication;
-	public static volatile JSONArray jsonArray = new JSONArray();
 
+	private List<Map<String, Object>> jsonArray = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java:49]: As you say, I suspect we will have much bigger performance issues if we start processing such large numbers of records.

However, `LinkedList` is probably appropriate since we would never expect to perform random access on this list.

File: README.md [hunk 6]
```diff
@@ -29,16 +29,22 @@ the [Kafka Connector documentation](http://docs.confluent.io/3.0.1/connect/userg
...
-Parameter | Value
----:|:---
-key.converter|org.apache.kafka.connect.json.JsonConverter
-value.converter|org.apache.kafka.connect.json.JsonConverter
-key.converter.schemas.enable|true
-value.converter.schemas.enable|true
+### Converter configuration
 
-Assume these settings in a file `connect-standalone.properties` or `connect-distributed.properties`.
+Usually the kafka distribution defaults (`connect-(standalone|distributed).properties`) are as follows:
```
[the remark this answers is not in the document]
Comment (tomblench) [reply README.md]: Done in ac8b55d

File: src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java [hunk 7]
```diff
@@ -39,6 +39,8 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.Properties;
+import java.util.stream.Collector;
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/common/utils/JavaCloudantUtil.java]: Cleaned up imports in da856f7

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java]: Used LinkedList in 7585eb7

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java [hunk 8]
```diff
@@ -61,44 +60,33 @@ public String version() {
...
-				
-				//OPTION C (not implemented): generate new id with  cloudant 
-				/*else {
-					LOG.info(MessageKey.GUID_SCHEMA + ": " + guid_schema);
-					LOG.warn(CloudantConst.CLOUDANT_DOC_ID + "from source database will removed");
-					
-					//remove Cloudant _id
-					jsonRecord.remove(CloudantConst.CLOUDANT_DOC_ID);
-				}*/
 			}
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java]: Done in 5583155 and removed test for it in 0231864. We can look at removing the option and documentation for it altogether in a later PR - I like the idea of having an SMT instead.

File: src/main/java/com/ibm/cloudant/kafka/connect/StructToMapConverter.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        } else {
+            Map<String, Object> toReturn = new HashMap<>();
+            if (record.value() instanceof Map) {
+                return convertMap((Map)record.value(), toReturn);
+            } else if (record.value() instanceof Struct) {
+                return convertStruct((Struct)record.value(), toReturn);    
+            }
+        }
+        throw new IllegalArgumentException(String.format("Type %s not supported, are you using value.converter=org.apache.kafka.connect.json.JsonConverter?", record.value().getClass()));
```
[the remark this answers is not in the document]
Comment (tomblench) [reply src/main/java/com/ibm/cloudant/kafka/connect/StructToMapConverter.java]: Adapted and used in 3fdee9d 

Note the first failure case:
```
                    throw new IllegalArgumentException(String.format("Type %s not supported with schema of type Map (or no schema)",
```
Because that leg covers both Map schema and no schema

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java [hunk 10]
```diff
@@ -48,65 +42,32 @@ public class CloudantSinkTask extends SinkTask {
...
-	
+		for (SinkRecord record : sinkRecords) {
+			Map<String, Object> recordValue = mapper.apply(record);
+			recordValue.remove(CloudantConst.CLOUDANT_REV);
+			jsonArray.add(recordValue);
+			if (jsonArray.size() >= batch_size) {
 				flush(null);
-	
 			} 
 		}
```
Comment (ricellis) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSinkTask.java:68]: Not essential, but now that the `mapper` is a functional interface and the code around here is more streamlined we could potentially use streams to take advantage of parallelism for the "mapping" part of the loop before sequentially processing the add/flush e.g. something like:
```java
		sinkRecords.stream()
			.map(mapper) // Convert ConnectRecord to Map
			.map(recordValueAsMap -> {
				recordValueAsMap.remove(CloudantConst.CLOUDANT_REV); // Remove the _rev
				return recordValueAsMap;
			})
			.sequential() // Avoid concurrent access to jsonArray
			.forEach(recordValueAsMap -> {
				jsonArray.add(recordValueAsMap);
				if (jsonArray.size() >= batch_size) {
					flush(null);
				} 
			});
```

**Review by emlaver — APPROVED** [review]
Looks good!
